### PR TITLE
Fix comma position before

### PR DESCRIPTION
**Pull request by d3x42**
Fix adding additional line before lines started with ',' when comma position is "before".
Fix treating ',' in the end of comment as lines separator

### CODE REVIEW

File: test/features/comments.ts [hunk 1]
```diff
@@ -274,4 +274,25 @@ export default function supportsComments(format: FormatFn, opts: CommentsConfig
       `);
     });
   }
+
+  it('should handle comma in comments with comaPosition:before', () => {
```
Comment (nene) [inline test/features/comments.ts:278]: This test should be in [test/options/commaPosition.ts](https://github.com/sql-formatter-org/sql-formatter/blob/master/test/options/commaPosition.ts) file.

File: src/formatter/formatCommaPositions.ts [hunk 2]
```diff
@@ -52,7 +52,7 @@ function groupCommaDelimitedLines(lines: string[]): string[][] {
     // when line ends with comma,
     // gather together all following lines that also end with comma,
     // plus one (which doesn't end with comma)
-    while (lines[i].match(/.*,(\s*(--.*)?$)/)) {
+    while (lines[i].match(/^[^--]*,(\s*(--.*)?$)/)) {
```
[the remark this answers is not in the document]
Comment (nene) [reply src/formatter/formatCommaPositions.ts:55]: I suggest you also add a test that checks what happens when minus operator is used, like in here:

```
SELECT
  5 - 6 AS col1, -- hello
  8 AS col2
```

**Review by nene — CHANGES_REQUESTED** [review]
Thanks for the PR. This seems to improve things, but there are some problems I've outlined in the comments.

You should also be aware that this whole `commaPosition:before` feature is fundamentally broken. For example, consider the scenario where `,` or `--` appears inside quotes. Just FYI. I'm not expecting you to fix it, as it would need to be completely rearchitected to fix these fundamental problems.

File: src/formatter/formatCommaPositions.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (d3x42) [reply src/formatter/formatCommaPositions.ts:55]: Hm, my bad. I'll fix it

[the remark this answers is not in the document]
Comment (d3x42) [reply src/formatter/formatCommaPositions.ts]: @nene I'm rewrited some methods in formatCommaPossition.ts to make commaPosition:before more stable.
Now query splits into lines with following format:

```
export interface Line {
  originalContent: string;
  commentAtLineStart: string;
  content: string;
  commentAtLineEnd: string;
}
```
Now all comments ranges detected before preparing lines. Then using those ranges new method splitContentAndComments prepare array of Lines.

File: test/options/commaPosition.ts [hunk 3]
```diff
@@ -1,12 +1,16 @@
-import { expect } from '@jest/globals';
-import dedent from 'dedent-js';
+import { expect } from "@jest/globals";
+import dedent from "dedent-js";
```
[the remark this answers is not in the document]
Comment (d3x42) [reply test/options/commaPosition.ts:2]: Sorry about that, seems like I'm accidentally run prettier. I'll change styles back

**Comment by nene** [comment]
I see you've added quite a lot of code to deal with some additional cases. I guess primarily to deal with multi-line comments.

This wasn't really a nice and pretty code to start with, and it's now grown quite a bit more complex. On top of that, as I pointed out earlier, this whole feature is really fundamentally broken. I haven't been interested enough in this feature to properly rearchitect it. I've actually seriously considered just completely dropping this, as it's just so broken.

I'm open to doing some small patch on the existing implementation when it improves it with some little amount of code, but I'd like to avoid any large attempts to "fix" it, because this whole feature is implemented in a completely wrong way. And unfortunately your work on this (which I'm sure is all very well-intended) just takes us further down that wrong path.

I haven't yet fully read your code, just slightly glanced at it. But I wanted to clearly lay out my expectations here. It's not really your fault that you stumbled upon this very broken part of the codebase. I'm most likely not going to accept this whole PR... though I might cherry-pick some smaller parts out of it.

**Comment by nene** [comment]
Upon further inspection I've decided that I'm not going to merge any parts of this pull request.

1. It doesn't properly fix the fundamental issues with this feature.
2. The fixes in here just complicate the already complex code (which I frankly barely understand by myself).
3. I'm going to drop this whole feature in the next major version.

If you really want to use this feature, you can simply take this whole `formatCommaPositions()` function and run it on top of the output from sql-formatter.

**Comment by d3x42** [comment]
Thanks, I'll take  ```formatCommaPositions()``` to product code

**Pull request closed without merging**